Perform some consistency checks in polyhedron primitive

tturpin opened the pull request:
This is a proposal to check for several geometric issues in the polyhedron primitive and output warnings which refer to the arguments. It follows my comments about this on #2351 and in [the forum](http://forum.openscad.org/User-Poll-What-do-you-want-to-see-from-OpenSCAD-development-td27802i60.html).

The implementation may not be of the best quality, but it demonstrates the idea. Here are the checks that are performed and cause an error or warning:
- the points must be vectors of 3 finite doubles (check unchanged)
- indices in the points vector must be in range (this was previously included in the "Unable to convert point ..." check)
- faces must have at least 3 vertices
- faces must not have consecutive identical vertices
- faces must not self-intersect
- polyhedron must be 2-manifold (this check is only an approximation, it uses the edge dictionary and will miss cases where 2n faces meet on the same edge, half of them with each half-edge).
- duplicate vertices are checked, I'm not sure if there should be a warning (this is explicitly supported in the manual)

Some other issues are not checked, most notably intersecting faces (other than self-intersection) and planarity of faces. But the implemented checks (unconnected edges in particular) give a sensible explanation for: #2351, #2398,  #2616, #2847, #2873, #3023. Unconnected edges are also reported for #2395, #2412 but in those two cases this is not the only issue.

I added a few cases to polyhedron-tests.scad, and looked at the messages ; other tests already trigger some warnings too (issue802.scad, line 109 for duplicate vertices, issue1246.scad, primitive-inf-tests.scad...)

I tried to keep the behavior unchanged, other than the messages. One exception is that invalid vertices (not a vector of 3 finite doubles) are now checked before starting to build the polyhedron, so the empty polyhedron is return in this case (previously the function returned whatever had already been inserted in the polyhedron). The returned result in case of errors warnings could deserve more reflection to either be more strict or try to "fix" some issues better (I'm not qualified for that as I don't know how robust CGAL is to various issues).

Let me know if the idea sounds reasonable !

thehans commented:
You updated the test case but it doesn't look like the expected result images are actually different, so I would just revert those pictures to previous version if that's true.

Also if the changes are just intended to add warnings, then you should probably edit tests/CMakeLists.txt to have that test included with the other echotests

tturpin commented:
Thank you for the hint about echotests, I hadn't found the way to test output. This made me realize that the new warnings / errors did not appear in the echo output, and were not properly handled in the GUI, as I implemented them in PrimitiveNode::createGeometry and not in PrimitiveModule::instantiate, where most errors and warnings are. I have submitted a new version with all checks in instantiate, and with echotests updated (and polyhedron-tests added). I also reverted the image tests.
Please note that I don't really know C++, so there are probably many mistakes in my code, especially about memory which I don't understand at all ;-)

tturpin commented:
Oups, this second commit has a serious bug, I'm currently running all tests before submitting again.
